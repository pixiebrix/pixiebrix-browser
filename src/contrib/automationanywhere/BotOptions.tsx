/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useMemo } from "react";
import { isEmpty, partial } from "lodash";
import { type BlockOptionProps } from "@/components/fields/schemaFields/genericOptionsFactory";
import {
  COMMON_PROPERTIES,
  ENTERPRISE_EDITION_COMMON_PROPERTIES,
} from "@/contrib/automationanywhere/RunBot";
import { type Schema } from "@/types/schemaTypes";
import { useField } from "formik";
import { useAsyncState } from "@/hooks/common";
import useDependency from "@/services/useDependency";
import ChildObjectField from "@/components/fields/schemaFields/ChildObjectField";
import ConnectedFieldTemplate from "@/components/form/ConnectedFieldTemplate";
import RemoteSelectWidget from "@/components/form/widgets/RemoteSelectWidget";
import RequireServiceConfig from "@/contrib/RequireServiceConfig";
import {
  cachedFetchBotFile,
  cachedSearchBots,
  cachedFetchDevicePools,
  cachedFetchDevices,
  cachedFetchFolder,
  cachedFetchRunAsUsers,
  cachedFetchSchema,
} from "@/contrib/automationanywhere/aaApi";
import { AUTOMATION_ANYWHERE_SERVICE_ID, type WorkspaceType } from "./contract";
import { isCommunityControlRoom } from "@/contrib/automationanywhere/aaUtils";
import BooleanWidget from "@/components/fields/schemaFields/widgets/BooleanWidget";
import RemoteMultiSelectWidget from "@/components/form/widgets/RemoteMultiSelectWidget";
import SelectWidget from "@/components/form/widgets/SelectWidget";
import { useAsyncEffect } from "use-async-effect";
import SchemaField from "@/components/fields/schemaFields/SchemaField";
import { Alert } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faInfoCircle } from "@fortawesome/free-solid-svg-icons";
import AsyncRemoteSelectWidget from "@/components/form/widgets/AsyncRemoteSelectWidget";
import { joinName } from "@/utils/formUtils";

const WORKSPACE_OPTIONS = [
  { value: "public", label: "Public" },
  { value: "private", label: "Private/Local" },
];

const BotLoadingMessage: React.FC = () => <span>Searching bots...</span>;
const BotNoOptionsMessage: React.FC = () => (
  <span>No bots found for query...</span>
);

const BotOptions: React.FunctionComponent<BlockOptionProps> = ({
  name,
  configKey,
}) => {
  const configName = partial(joinName, name, configKey);

  const { hasPermissions, config } = useDependency(
    AUTOMATION_ANYWHERE_SERVICE_ID
  );

  const [{ value: workspaceType }, , { setValue: setWorkspaceType }] =
    useField<string>(configName("workspaceType"));

  const [{ value: fileId }] = useField<string>(configName("fileId"));

  const [{ value: isAttended = false }] = useField<boolean>(
    configName("isAttended")
  );

  const [{ value: awaitResult }] = useField<boolean | null>(
    configName("awaitResult")
  );

  // Default the workspaceType based on the file id
  useAsyncEffect(
    async (isMounted) => {
      if (config && isCommunityControlRoom(config.config.controlRoomUrl)) {
        // In community edition, each user just works in their own private workspace
        await setWorkspaceType("private");
      }

      // `workspaceType` is optional because it's not required to run the bot. However, we need it to populate dropdowns
      // for the fields in the fieldset
      if (hasPermissions && config && workspaceType == null && fileId) {
        const result = await cachedFetchBotFile(config, fileId);
        const workspaceType =
          result.workspaceType === "PUBLIC" ? "public" : "private";
        if (isMounted()) {
          await setWorkspaceType(workspaceType);
        }
      }

      // Leave setWorkspaceType off the dependency list because Formik changes reference on each render
    },
    [config, fileId, hasPermissions, workspaceType]
  );

  const [remoteSchema, remoteSchemaPending, remoteSchemaError] =
    useAsyncState(async () => {
      if (hasPermissions && config && fileId) {
        return cachedFetchSchema(config, fileId);
      }

      return null;
    }, [config, fileId, hasPermissions]);

  // Don't care about pending/error state b/c we just fall back to displaying the folderId
  const [folder] = useAsyncState(async () => {
    if (hasPermissions && config && config.config.folderId) {
      return cachedFetchFolder(config, config.config.folderId);
    }

    return null;
  }, [config, hasPermissions]);

  // Additional args passed to the remote options factories
  const factoryArgs = useMemo(
    () => ({
      // Default to "private" because that's compatible with both CE and EE
      // The workspaceType can be temporarily null when switching between CR configurations
      workspaceType: (workspaceType as WorkspaceType) ?? "private",
    }),
    [workspaceType]
  );

  return (
    <RequireServiceConfig
      serviceSchema={COMMON_PROPERTIES.service as Schema}
      serviceFieldName={configName("service")}
    >
      {({ config }) => (
        <>
          {!isCommunityControlRoom(config.config.controlRoomUrl) && (
            <ConnectedFieldTemplate
              label="Workspace"
              name={configName("workspaceType")}
              description="The Control Room Workspace"
              as={SelectWidget}
              defaultValue="private"
              options={WORKSPACE_OPTIONS}
            />
          )}

          {!isEmpty(config.config.folderId) && (
            <Alert variant="info">
              <FontAwesomeIcon icon={faInfoCircle} /> Displaying available bots
              from folder{" "}
              {folder?.name
                ? `'${folder.name}' (${config.config.folderId})`
                : config.config.folderId}{" "}
              configured on the integration. To choose from all bots in the
              workspace, remove the folder from the integration configuration.
            </Alert>
          )}

          {
            // Use AsyncRemoteSelectWidget instead of RemoteSelectWidget because the former can handle
            // Control Rooms with lots of bots
            // https://github.com/pixiebrix/pixiebrix-extension/issues/5260
            <ConnectedFieldTemplate
              label="Bot"
              name={configName("fileId")}
              description="The Automation Anywhere bot to run. Type a query to search available bots by name"
              as={AsyncRemoteSelectWidget}
              defaultOptions
              // Ensure we get current results, because there's not refresh button
              cacheOptions={false}
              optionsFactory={cachedSearchBots}
              loadingMessage={BotLoadingMessage}
              noOptonsMessage={BotNoOptionsMessage}
              factoryArgs={factoryArgs}
              config={config}
            />
          }

          {isCommunityControlRoom(config.config.controlRoomUrl) ? (
            <ConnectedFieldTemplate
              label="Device"
              name={configName("deviceId")}
              description="The device to run the bot on"
              as={RemoteSelectWidget}
              optionsFactory={cachedFetchDevices}
              factoryArgs={factoryArgs}
              config={config}
            />
          ) : (
            <>
              {workspaceType === "public" && (
                <>
                  {isAttended && (
                    <Alert variant="info">
                      <FontAwesomeIcon icon={faInfoCircle} /> In attended mode,
                      the bot will run using the authenticated user&apos;s
                      credentials. You will not be able to run the bot using a
                      Bot Creator license. Switch to unattended mode to test
                      using a Bot Creator license.
                    </Alert>
                  )}
                  <ConnectedFieldTemplate
                    label="Attended"
                    name={configName("isAttended")}
                    description="Run the bot in attended mode, using the authenticated user's device. Requires an Attended Bot license"
                    as={BooleanWidget}
                  />
                  {!isAttended && (
                    <>
                      <ConnectedFieldTemplate
                        label="Run as Users"
                        name={configName("runAsUserIds")}
                        description="The user(s) to run the bots"
                        as={RemoteMultiSelectWidget}
                        optionsFactory={cachedFetchRunAsUsers}
                        factoryArgs={factoryArgs}
                        blankValue={[]}
                        config={config}
                      />
                      <ConnectedFieldTemplate
                        label="Device Pools"
                        name={configName("poolIds")}
                        description="A device pool that has at least one active device (optional)"
                        as={RemoteMultiSelectWidget}
                        optionsFactory={cachedFetchDevicePools}
                        factoryArgs={factoryArgs}
                        blankValue={[]}
                        config={config}
                      />
                    </>
                  )}
                </>
              )}
              <ConnectedFieldTemplate
                label="Await Result"
                name={configName("awaitResult")}
                description="Wait for the bot to run and return the output"
                as={BooleanWidget}
              />
              {awaitResult && (
                <SchemaField
                  label="Result Timeout (Milliseconds)"
                  name={configName("maxWaitMillis")}
                  schema={
                    ENTERPRISE_EDITION_COMMON_PROPERTIES.maxWaitMillis as Schema
                  }
                  // Mark as required so the widget defaults to showing the number entry
                  isRequired
                />
              )}
            </>
          )}

          {fileId && (
            <ChildObjectField
              heading="Input Arguments"
              name={configName("data")}
              schema={remoteSchema}
              schemaError={remoteSchemaError}
              schemaLoading={remoteSchemaPending}
              isRequired
            />
          )}
        </>
      )}
    </RequireServiceConfig>
  );
};

export default BotOptions;
