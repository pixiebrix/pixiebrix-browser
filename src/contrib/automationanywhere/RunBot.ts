/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Transformer, type UnknownObject } from "@/types";
import {
  type BlockArg,
  type BlockOptions,
  type Schema,
  type SchemaProperties,
} from "@/core";
import { validateRegistryId } from "@/types/helpers";
import { isCommunityControlRoom } from "@/contrib/automationanywhere/aaUtils";
import {
  DEFAULT_MAX_WAIT_MILLIS,
  pollEnterpriseResult,
  runCommunityBot,
  runEnterpriseBot,
} from "@/contrib/automationanywhere/aaApi";
import {
  type BotArgs,
  type EnterpriseBotArgs,
} from "@/contrib/automationanywhere/aaTypes";
import { BusinessError, PropError } from "@/errors/businessErrors";
import {
  CONTROL_ROOM_OAUTH_SERVICE_ID,
  CONTROL_ROOM_SERVICE_ID,
} from "@/services/constants";
import { isEmpty } from "lodash";
import { getCachedAuthData } from "@/background/messenger/api";

export const AUTOMATION_ANYWHERE_RUN_BOT_ID = validateRegistryId(
  "@pixiebrix/automation-anywhere/run-bot"
);

export const COMMON_PROPERTIES: SchemaProperties = {
  service: {
    anyOf: [CONTROL_ROOM_SERVICE_ID, CONTROL_ROOM_OAUTH_SERVICE_ID].map(
      (id) => ({
        $ref: `https://app.pixiebrix.com/schemas/services/${id}`,
      })
    ),
  },
  workspaceType: {
    type: "string",
    enum: ["private", "public"],
    description: "The workspace that contains the bot",
    default: "private",
  },
  fileId: {
    type: "string",
    description: "The file id of the bot",
    format: "\\d+",
  },
  data: {
    type: "object",
    description: "The input data for the bot",
    additionalProperties: true,
  },
};

const COMMUNITY_EDITION_PROPERTIES: SchemaProperties = {
  deviceId: {
    type: "string",
    description: "The device to run the bot",
    format: "\\d+",
  },
};

export const ENTERPRISE_EDITION_COMMON_PROPERTIES: SchemaProperties = {
  awaitResult: {
    type: "boolean",
    default: false,
    description: "Wait for the bot to complete and return the output",
  },
  maxWaitMillis: {
    type: "number",
    default: DEFAULT_MAX_WAIT_MILLIS,
    description:
      "Maximum time (in milliseconds) to wait for the bot to complete when awaiting result.",
  },
};

export const ENTERPRISE_EDITION_PUBLIC_PROPERTIES: SchemaProperties = {
  runAsUserIds: {
    type: "array",
    description: "The user(s) to run the bot",
    items: {
      type: "string",
    },
  },
  poolIds: {
    type: "array",
    description: "A device pool that has at least one active device (optional)",
    items: {
      type: "string",
    },
  },
};

export class RunBot extends Transformer {
  constructor() {
    super(
      AUTOMATION_ANYWHERE_RUN_BOT_ID,
      "Run Automation Anywhere Bot",
      "Run an Automation Anywhere Bot via the Control Room API"
    );
  }

  inputSchema: Schema = {
    $schema: "https://json-schema.org/draft/2019-09/schema#",
    anyOf: [
      {
        type: "object",
        properties: {
          ...COMMON_PROPERTIES,
          ...COMMUNITY_EDITION_PROPERTIES,
        },
      },
      {
        type: "object",
        properties: {
          ...COMMON_PROPERTIES,
          ...ENTERPRISE_EDITION_COMMON_PROPERTIES,
          ...ENTERPRISE_EDITION_PUBLIC_PROPERTIES,
        },
      },
      {
        type: "object",
        properties: {
          ...COMMON_PROPERTIES,
          ...ENTERPRISE_EDITION_COMMON_PROPERTIES,
        },
      },
    ],
  };

  defaultOutputKey = "bot";

  override outputSchema: Schema = {
    $schema: "https://json-schema.org/draft/2019-09/schema#",
    type: "object",
    additionalProperties: true,
  };

  async transform(
    args: BlockArg<BotArgs>,
    { logger }: BlockOptions
  ): Promise<UnknownObject> {
    const {
      awaitResult,
      maxWaitMillis = DEFAULT_MAX_WAIT_MILLIS,
      service,
      workspaceType,
    } = args;

    if (isCommunityControlRoom(service.config.controlRoomUrl)) {
      if (!("deviceId" in args)) {
        throw new PropError(
          "deviceId is required for Community Edition",
          this.id,
          "deviceId",
          undefined
        );
      }

      if (awaitResult) {
        throw new PropError(
          "Cannot await result with Community Edition",
          this.id,
          "awaitResult",
          awaitResult
        );
      }

      await runCommunityBot(args);
      return {};
    }

    const enterpriseBotArgs: EnterpriseBotArgs =
      args as unknown as EnterpriseBotArgs;

    let runAsUserIds: number[];
    if (
      workspaceType === "public" ||
      !isEmpty(enterpriseBotArgs.runAsUserIds)
    ) {
      runAsUserIds = enterpriseBotArgs.runAsUserIds;
    } else if (
      workspaceType === "private" &&
      service.serviceId === CONTROL_ROOM_OAUTH_SERVICE_ID
    ) {
      throw new PropError(
        "Running local bots with OAuth2 authentication is not yet supported",
        this.id,
        "workspaceType",
        workspaceType
      );
    } else if (workspaceType === "private" || workspaceType == null) {
      // Get the user id from the cached token data. AA doesn't have any endpoints for retrieving the user id that
      // we could automatically fetch in the Page Editor
      const userData = (await getCachedAuthData(service.id)) as unknown as {
        user: { id: number };
      };
      if (!userData) {
        throw new BusinessError(
          "User profile for Control Room not found. Reconnect the Control Room integration"
        );
      }

      const userId = userData?.user?.id;
      if (userId == null) {
        // Indicates the Control Room provided an unexpected response shape
        throw new Error("Control Room token response missing user id.");
      }

      runAsUserIds = [userId];
    }

    const deployment = await runEnterpriseBot({
      ...enterpriseBotArgs,
      runAsUserIds,
    });

    if (!awaitResult) {
      return deployment;
    }

    return pollEnterpriseResult({
      service,
      deploymentId: deployment.deploymentId,
      logger,
      maxWaitMillis,
    });
  }
}
