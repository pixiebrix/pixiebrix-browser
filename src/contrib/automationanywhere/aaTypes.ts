/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type SanitizedIntegrationConfig } from "@/types/integrationTypes";
import { type UnknownObject } from "@/types/objectTypes";

export type CommunityBotArgs = {
  service: SanitizedIntegrationConfig;
  fileId: string;
  deviceId: string;
  data: UnknownObject;
};

export type EnterpriseBotArgs = {
  service: SanitizedIntegrationConfig;
  fileId: string;

  /**
   * Run the bot in attended mode using the authenticated user's device.
   * @since 1.7.16
   */
  isAttended?: boolean;

  runAsUserIds: number[];
  poolIds: string[];
  data: UnknownObject;
};

export type BotArgs = (CommunityBotArgs | EnterpriseBotArgs) & {
  workspaceType?: "public" | "private" | null;
  awaitResult: boolean;
  maxWaitMillis: number;
};
