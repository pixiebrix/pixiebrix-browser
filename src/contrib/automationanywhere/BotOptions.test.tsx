/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { menuItemFormStateFactory } from "@/testUtils/factories";
import { type FormState } from "@/pageEditor/extensionPoints/formStateTypes";
import { render } from "@/extensionConsole/testHelpers";
// eslint-disable-next-line no-restricted-imports -- TODO: Fix over time
import { Formik } from "formik";
import { CONTROL_ROOM_SERVICE_ID } from "@/services/constants";
import { AUTOMATION_ANYWHERE_RUN_BOT_ID } from "@/contrib/automationanywhere/RunBot";
import BotOptions from "@/contrib/automationanywhere/BotOptions";
import useDependency from "@/services/useDependency";
import { makeVariableExpression } from "@/runtime/expressionCreators";
import { uuidv4 } from "@/types/helpers";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";
import { type OutputKey } from "@/types/runtimeTypes";
import { type IService } from "@/types/serviceTypes";

jest.mock("webext-detect-page", () => ({
  isDevToolsPage: () => true,
  isExtensionContext: () => true,
  isBackground: () => false,
  isContentScript: () => false,
}));

jest.mock("@/services/useDependency", () =>
  jest.fn().mockReturnValue({
    // Pass minimal arguments
    config: undefined,
    service: undefined,
    hasPermissions: true,
    requestPermissions: jest.fn(),
  })
);
jest.mock("@/hooks/auth", () => ({
  useAuthOptions: jest.fn().mockReturnValue([[], jest.fn()]),
}));
jest.mock("@/hooks/auth");
jest.mock("@/contentScript/messenger/api");

function makeBaseState() {
  const baseFormState = menuItemFormStateFactory();
  baseFormState.services = [
    {
      id: CONTROL_ROOM_SERVICE_ID,
      outputKey: "automationAnywhere" as OutputKey,
    },
  ];
  baseFormState.extension.blockPipeline = [
    {
      id: AUTOMATION_ANYWHERE_RUN_BOT_ID,
      config: {
        service: null,
        fileId: null,
        data: {},
      },
    },
  ];
  return baseFormState;
}

function renderOptions(formState: FormState = makeBaseState()) {
  return render(
    <Formik onSubmit={jest.fn()} initialValues={formState}>
      <BotOptions name="extension.blockPipeline.0" configKey="config" />
    </Formik>
  );
}

beforeAll(() => {
  registerDefaultWidgets();
});

describe("BotOptions", () => {
  it("should require selected integration", async () => {
    const result = renderOptions();

    expect(result.queryByText("Workspace")).toBeNull();
    expect(result.container).toMatchSnapshot();
  });

  it("should render default enterprise fields for private workspace", async () => {
    (useDependency as jest.Mock).mockReturnValue({
      config: {
        id: uuidv4(),
        config: {
          controlRoomUrl: "https://control.room.com",
        },
      },
      service: {} as IService,
      hasPermissions: true,
      requestPermissions: jest.fn(),
    });

    const base = makeBaseState();
    base.extension.blockPipeline[0].config.service = makeVariableExpression(
      "@automationAnywhere"
    );

    const result = renderOptions(base);

    expect(result.queryByText("Workspace")).not.toBeNull();
    expect(result.queryByText("Bot")).not.toBeNull();
    expect(result.queryByText("Run as Users")).toBeNull();
    expect(result.queryByText("Device Pools")).toBeNull();
    expect(result.queryByText("Await Result")).not.toBeNull();
    expect(result.queryByText("Result Timeout (Milliseconds)")).toBeNull();

    // There's non-determinism in React Selects ids: react-select-X-live-region
    // expect(result.container).toMatchSnapshot();
  });

  it("should render community fields", async () => {
    (useDependency as jest.Mock).mockReturnValue({
      config: {
        id: uuidv4(),
        config: {
          controlRoomUrl:
            "https://community2.cloud-2.automationanywhere.digital",
        },
      },
      service: {} as IService,
      hasPermissions: true,
      requestPermissions: jest.fn(),
    });

    const base = makeBaseState();
    base.extension.blockPipeline[0].config.service = makeVariableExpression(
      "@automationAnywhere"
    );

    const result = renderOptions(base);

    // Community only supports private workspace, so we don't show the field
    expect(result.queryByText("Workspace")).toBeNull();
    expect(result.queryByText("Bot")).not.toBeNull();
    expect(result.queryByText("Run as Users")).toBeNull();
    expect(result.queryByText("Device Pools")).toBeNull();
    expect(result.queryByText("Await Result")).toBeNull();
    expect(result.queryByText("Result Timeout (Milliseconds)")).toBeNull();

    // There's non-determinism in React Selects ids: react-select-X-live-region
    // expect(result.container).toMatchSnapshot();
  });

  it("should render default enterprise fields for public workspace", async () => {
    (useDependency as jest.Mock).mockReturnValue({
      config: {
        id: uuidv4(),
        config: {
          controlRoomUrl: "https://control.room.com",
        },
      },
      service: {} as IService,
      hasPermissions: true,
      requestPermissions: jest.fn(),
    });

    const base = makeBaseState();
    base.extension.blockPipeline[0].config.workspaceType = "public";
    base.extension.blockPipeline[0].config.service = makeVariableExpression(
      "@automationAnywhere"
    );

    const result = renderOptions(base);

    expect(result.queryByText("Workspace")).not.toBeNull();
    expect(result.queryByText("Bot")).not.toBeNull();
    expect(result.queryByText("Attended")).not.toBeNull();
    expect(result.queryByText("Run as Users")).not.toBeNull();
    expect(result.queryByText("Device Pools")).not.toBeNull();
    expect(result.queryByText("Await Result")).not.toBeNull();
    expect(result.queryByText("Result Timeout (Milliseconds)")).toBeNull();

    // There's non-determinism in React Selects ids: react-select-X-live-region
    // expect(result.container).toMatchSnapshot();
  });

  it("should render attended enterprise fields for public workspace", async () => {
    (useDependency as jest.Mock).mockReturnValue({
      config: {
        id: uuidv4(),
        config: {
          controlRoomUrl: "https://control.room.com",
        },
      },
      service: {} as IService,
      hasPermissions: true,
      requestPermissions: jest.fn(),
    });

    const base = makeBaseState();
    base.extension.blockPipeline[0].config.workspaceType = "public";
    base.extension.blockPipeline[0].config.isAttended = true;
    base.extension.blockPipeline[0].config.service = makeVariableExpression(
      "@automationAnywhere"
    );

    const result = renderOptions(base);

    expect(result.queryByText("Workspace")).not.toBeNull();
    expect(result.queryByText("Bot")).not.toBeNull();
    expect(result.queryByText("Attended")).not.toBeNull();
    expect(result.queryByText("Run as Users")).toBeNull();
    expect(result.queryByText("Device Pools")).toBeNull();
    expect(result.queryByText("Await Result")).not.toBeNull();
    expect(result.queryByText("Result Timeout (Milliseconds)")).toBeNull();

    // There's non-determinism in React Selects ids: react-select-X-live-region
    // expect(result.container).toMatchSnapshot();
  });

  it("should render result timeout", async () => {
    (useDependency as jest.Mock).mockReturnValue({
      config: {
        id: uuidv4(),
        config: {
          controlRoomUrl: "https://control.room.com",
        },
      },
      service: {} as IService,
      hasPermissions: true,
      requestPermissions: jest.fn(),
    });

    const base = makeBaseState();
    base.extension.blockPipeline[0].config.awaitResult = true;
    base.extension.blockPipeline[0].config.service = makeVariableExpression(
      "@automationAnywhere"
    );

    const result = renderOptions(base);

    expect(result.queryByText("Await Result")).not.toBeNull();
    expect(result.queryByText("Result Timeout (Milliseconds)")).not.toBeNull();
  });
});
