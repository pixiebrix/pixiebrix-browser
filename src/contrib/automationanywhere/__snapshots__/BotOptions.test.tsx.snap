// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BotOptions should require selected integration 1`] = `
<div>
  <div>
    <div
      class="formGroup form-group row"
    >
      <label
        class="label form-label col-form-label col-xl-2 col-lg-3"
        for="extension.blockPipeline.0.config.service"
      >
        Integration
      </label>
      <div
        class="col-xl-10 col-lg-9"
      >
        <div
          class="d-flex"
          data-widget="ServiceSelectWidget"
        >
          <div
            class="flex-grow-1"
          >
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class=" css-1s2u09g-control"
              >
                <div
                  class=" css-319lph-ValueContainer"
                >
                  <div
                    class=" css-14el2xx-placeholder"
                    id="react-select-2-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    class=" css-6j8wv5-Input"
                    data-value=""
                  >
                    <input
                      aria-autocomplete="list"
                      aria-describedby="react-select-2-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="extension.blockPipeline.0.config.service"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <input
                name="extension.blockPipeline.0.config.service"
                type="hidden"
                value=""
              />
            </div>
          </div>
          <div>
            <button
              class="btn btn-info"
              title="Refresh configured integrations"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-sync fa-w-16 "
                data-icon="sync"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M440.65 12.57l4 82.77A247.16 247.16 0 0 0 255.83 8C134.73 8 33.91 94.92 12.29 209.82A12 12 0 0 0 24.09 224h49.05a12 12 0 0 0 11.67-9.26 175.91 175.91 0 0 1 317-56.94l-101.46-4.86a12 12 0 0 0-12.57 12v47.41a12 12 0 0 0 12 12H500a12 12 0 0 0 12-12V12a12 12 0 0 0-12-12h-47.37a12 12 0 0 0-11.98 12.57zM255.83 432a175.61 175.61 0 0 1-146-77.8l101.8 4.87a12 12 0 0 0 12.57-12v-47.4a12 12 0 0 0-12-12H12a12 12 0 0 0-12 12V500a12 12 0 0 0 12 12h47.35a12 12 0 0 0 12-12.6l-4.15-82.57A247.17 247.17 0 0 0 255.83 504c121.11 0 221.93-86.92 243.55-201.82a12 12 0 0 0-11.8-14.18h-49.05a12 12 0 0 0-11.67 9.26A175.86 175.86 0 0 1 255.83 432z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
        <small
          class="text-muted form-text"
        >
          <span>
            Select an integration configuration.
             
            <a
              href="chrome-extension://abcxyz/options.html#/services"
              rel="noopener noreferrer"
              target="_blank"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-cloud fa-w-20 "
                data-icon="cloud"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 640 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M537.6 226.6c4.1-10.7 6.4-22.4 6.4-34.6 0-53-43-96-96-96-19.7 0-38.1 6-53.3 16.2C367 64.2 315.3 32 256 32c-88.4 0-160 71.6-160 160 0 2.7.1 5.4.2 8.1C40.2 219.8 0 273.2 0 336c0 79.5 64.5 144 144 144h368c70.7 0 128-57.3 128-128 0-61.9-44-113.6-102.4-125.4z"
                  fill="currentColor"
                />
              </svg>
              Â Configure additional integrations here.
            </a>
          </span>
        </small>
      </div>
    </div>
  </div>
</div>
`;
